---
title: "Ptychodera Cisreg Development:Re-Analysis of Anneissia japonica"
author: "Alberto Perez-Posada"
date: "6/6/2023"
output: html_document
---

```{r setup, include=FALSE}
dir <- '/home/ska/aperpos/projects/ptychodera_cisreg_development/'
fcha <- function(){ gsub("-","",Sys.Date()) }

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = dir)
options(scipen=999)
Sys.setenv(VROOM_CONNECTION_SIZE=5000072)
```

## About

In this markdown, we will do a re-analysis of the transcriptome of *Anneissia japonica* (Li et al., 2020).

## Load libraries

```{r libraries, message=FALSE, warning=FALSE}
library(tidyverse)
library(stringr)
library(BiocGenerics)
library(EDASeq)
library(DESeq2)
library(tximport)
library(limma)
library(ComplexHeatmap)
library(circlize)
library(RColorBrewer)
library(colorspace)
```

## Load functions

```{r load_functions, message = FALSE}
source("code/r_code/r_functions/sourcefolder.R")

sourceFolder(
  "code/r_code/r_functions",
  recursive = TRUE
  )

sourceFolder(
  "code/r_code/r_general",
  recursive = TRUE
  )
```

## Load data

We will load a sample table with all the information of each developmental transcriptome, in a similar way we did with Ptychodera.

```{r}
samplesdir <- "/home/ska/aperpos/Def_Pfla/other_spp/echinodermata/crinoidea/anneissia_japonica/outputs/RNASeq_anneissia/devstages/kallisto_20210621/"
  # "outputs/comparative/reanalysis/blan/kallisto/abundances/"

sampleTable <- 
  read.table(
    "/home/ska/aperpos/Def_Pfla/other_spp/echinodermata/crinoidea/anneissia_japonica/outputs/RNASeq_anneissia/devstages/kallisto_20210621/20230615_sample_stage_table_deseq2.txt",
    #"outputs/comparative/reanalysis/blan/kallisto/abundances/table_runs_samples_blan_RNA_2.tsv",
    header=F
  )

sampleTable <- 
  merge(
    sampleTable,
    as.data.frame(c(list.dirs(samplesdir,full.names = F)[-1])),
    by=1
    )

colnames(sampleTable) <- c("sample","condition","batch")

sampleTable <- sampleTable[order(sampleTable$condition,sampleTable$batch),]

sampleTable$batch <- 
  factor(sampleTable$batch, levels = unique(sort(sampleTable$batch)))

sampleTable$condition <- factor(sampleTable$condition)

rownames(sampleTable) <- sampleTable[,1]
sampleTable$sample <- NULL #sort the table properly and retrieve the conditions in developmental order. same for counts

ajap_condition_x <- sampleTable$condition
```

We will also load a dictionary between the gene ids of our mapping and the gene ids of the proteome used in comparative analyses (OMA, Orthofinder). This will be useful later in the future.

```{r load_dict}
#we have to translate all the IDs as soon as possible in the analysis
aj_aenja <- read.table("outputs/comparative/oma/dicts/AENJA.dct", col.names = c("key","value"))
```

## Tximport

We load the transcriptome mapping from kallisto here using the data of the sample table:

```{r tximport}
library(tximport)
files <- 
  list.files(samplesdir, "h5",recursive = T, full.names = T)

names(files) <- 
  list.dirs(samplesdir,full.names = F)[-1]

#put these file names in same order as sample table

files <- files[match(rownames(sampleTable),names(files))]

ajap_txi.kallisto <- 
  tximport(files, type = "kallisto", txOut = TRUE)
```

## DGE using DESeq2

Load the transcriptome counts in a DESeq2 object:

```{r deseq_object}
ajap_ddsTxi <- 
  DESeqDataSetFromTximport(
    ajap_txi.kallisto, 
    colData = sampleTable,
    design = ~ batch + condition
    )

rownames(ajap_ddsTxi) <- 
  translate_ids(
    x = rownames(ajap_ddsTxi),
    dict = aj_aenja
    )

```

Relevel the factor to first developmental stage.

```{r relevel}
#differential expression analysis
ajap_ddsTxi$condition <- 
  relevel(ajap_ddsTxi$condition, ref = c("01_Anj_UFegg"))
```

We estimate the size factors and keep the genes with more than 10 counts throughout the whole dataset. After this we do the actual DGE analysis using Likelihood-Ratio Test as the method because this is time-series data.

```{r DGE}
dds_ajap <- estimateSizeFactors(ajap_ddsTxi)

nc_ajap <- counts(dds_ajap, normalized=TRUE)
filter <- rowSums(nc_ajap >= 10) >= 2
dds_ajap <- dds_ajap[filter,]

dds_ajap <- 
  DESeq(
    dds_ajap,
    parallel = T,
    BPPARAM = BiocParallel::MulticoreParam(12),
    test = "LRT",
    reduced = ~batch
    )
```

We extract the results

```{r}
res_ajap <- results(dds_ajap)
res_ajap
```

And we keep the significant results (p.adj < 0.1)

```{r}
cutoff = 0.1 # try 0.05?
signif_ajap <- subset(res_ajap, padj<cutoff)
signif_ajap <- signif_ajap[order(signif_ajap$padj),]
```

We extract the variance-transformed data.

```{r datat ransform}
# raw counts
ajap_counts <- as.data.frame(counts(dds_ajap))
colnames(ajap_counts) <- 
  paste(sampleTable$condition,sampleTable$batch,sep="_")
ajap_counts <- ajap_counts[,order(colnames(ajap_counts))]

# vs-data
ajap_vsd <- as.data.frame(assay(vst(dds_ajap)))
colnames(ajap_vsd) <- 
  paste(sampleTable$condition,sampleTable$batch,sep="_")
ajap_vsd <- ajap_vsd[,order(colnames(ajap_vsd))]

#vs-data of stage-specific genes
rna_ajap <- ajap_vsd[rownames(ajap_vsd) %in% rownames(signif_ajap),]
```

We calculate rows by replicates:

```{r rowmeans_repl}
RNA_ajap <- 
  data.frame(
    base::sapply(
      levels(ajap_condition_x),
      rowMeans_by_repl,rna_ajap
      )
  )

colnames(RNA_ajap) <- sub("^X","",colnames(RNA_ajap))

```

## Correlation Plot

```{r cor, fig.width = 7, fig.height = 6}
ajap_cor <- 
  Heatmap(
    cor(ajap_vsd, method = "spearman"),
    name="Anneissia\n(Spearman)",
    cluster_rows = F,
    cluster_columns = F,
    col = colorRampPalette(rev(brewer.pal(7,"RdYlBu")))(100)
  )

draw(ajap_cor)
```

## PCA plots

```{r PCA,fig.height = 6, fig.width = 12}
par(mfrow = c(1,2))
plotPCA(as.matrix(rna_ajap), col=rainbow(20), cex=0.75,main="Genes diff. regulated\nduring development\n(separate replicates)") #fix colors
plotPCA(as.matrix(RNA_ajap),col=viridis::viridis(12),cex=0.75,main="Genes diff. regulated\nduring development\n(merged replicates)")
par(mfrow = c(1,1))
```

## Clustering using hierarchical clustering

We will group the stage-specific expressed genes in groups of stage-specific coexpression using hierarchical clustering.

```{r heatmap}
ajap_hc <- hclust(dist(t(scale(t(RNA_ajap)))), method = "ward.D2")
plot(ajap_hc,cex=0.3)
abline(50,0,col="red")

RNA_ajap$clusterID <- cutree(ajap_hc, h=50)
unique(RNA_ajap$clusterID)
```

```{r, fig.width = 4, fig.height = 7.5}
  Heatmap(
    t(scale(t(RNA_ajap[,1:16]))),
    name = "gene exp",
    # left_annotation= bl_col_annot,
    row_split=RNA_ajap$clusterID,
    show_row_names = F,
    cluster_rows=F,
    cluster_columns=F,
    # row_title=NULL,
    left_annotation = 
      rowAnnotation(
        cluster = anno_block(
          gp = gpar(col=NA,fill=NA),
          labels = unique(sort(RNA_ajap$newcl)), 
          labels_gp = gpar(col = "black", cex=0.7),
          labels_just = "right",
          labels_offset = unit(1, "npc"),
          labels_rot=0
          )
        ),
  column_names_side="top",
  col = diverging_hcl(20,"Purple-Brown")
)
```

Based on visual inspection, we assign new cluster names to give them new ordering:

```{r newcl}
ajap_newcl <- 
  read.table(
    file = "outputs/comparative/reanalysis/ajap/ajap_newcl.tsv",
    header = FALSE,
    col.names = c("clusterID","newcl")
  )

RNA_ajap$newcl <-
  translate_ids(
    x = RNA_ajap$clusterID,
    dict = ajap_newcl
  )

#We do not need the original hierachical cluster ids anymore
RNA_ajap$clusterID <- NULL

```

And a final Heatmap for visualisation:

```{r, fig.width = 4, fig.height = 7.5}
set.seed(1234)
RNA_ajap_viz <- RNA_ajap[sample(nrow(RNA_ajap),1000),]

aj_col_annot <- 
  rowAnnotation(
    sample = factor(RNA_ajap$newcl),
    show_annotation_name = FALSE
    )

ajap_heat <- 
  Heatmap(
    t(scale(t(RNA_ajap_viz[,1:16]))),
    name = "gene exp",
    # left_annotation= aj_col_annot,
    row_split=RNA_ajap_viz$newcl,
    show_row_names = F,
    cluster_rows=F,
    cluster_columns=F,
    row_title=NULL,
    left_annotation = 
      rowAnnotation(
        cluster = anno_block(
          gp = gpar(col=NA,fill=NA),
          labels = unique(sort(RNA_ajap_viz$newcl)), 
          labels_gp = gpar(col = "black", cex=0.7),
          labels_just = "right",
          labels_offset = unit(1, "npc"),
          labels_rot=0
          )
        ),
  column_names_side="top",
  col = diverging_hcl(20,"Purple-Brown")
)

draw(ajap_heat)
```

And the average expression profile of the different clusters of stage-specific genes:

```{r avg_profiles}
ajap_cl_avgs <-
  aggregate(
    t(scale(t(RNA_ajap[, 1:16]))),
    by = list(cluster = RNA_ajap$newcl),
    FUN = mean
  )[, -1]

ajap_cl_colors <- 
  colorRampPalette(colors=brewer.pal(11,"Spectral"))(16)

plot(
  0,
  xlim = c(1, 16),
  ylim = c(-2, 3),
  type = "n",
  main = "Average Expression Profile of\nStage-Specific Gene Clusters",
  xlab = "stage of development",
  ylab = "z-score Expression"
)
for (i in 1:nrow(ajap_cl_avgs)) {
  lines(c(1:16),
        ajap_cl_avgs[i,],
        type = "l",
        col = ajap_cl_colors[i],
        lwd = 2
  )
}
legend(
  x = 1,
  y = 3,
  legend = rownames(ajap_cl_avgs),
  fill = ajap_cl_colors,
  ncol = 5,
  cex = 0.6
)
```

```{r ajap_avg_heat, message = FALSE, warning = FALSE, fig.width = 4, fig.height = 5}
ajap_heat_avg <- 
  Heatmap(
    ajap_cl_avgs,
    name = "gene exp",
    row_labels = seq(nrow(ajap_cl_avgs)),
    cluster_rows=F,
    cluster_columns=F,
    column_names_side="top",
    col = diverging_hcl(20,"Purple-Brown")
  )

draw(ajap_heat_avg)
```
```{r}
pdf("graphics/ajap_clusters.pdf", height = 5, width = 4)
draw(ajap_heat_avg)
dev.off()
```


## Save the data

We finally create a table of gene <-> cluster for later use in our comparative analyses:

```{r ajap_cl}
ajap_cl <- 
  data.frame(
    id = rownames(RNA_ajap),
    cID = RNA_ajap$newcl
  )
```

And we save the data for TF analysis and for pairwise comparisons:

```{r save}
save(
  ajap_counts,
  ajap_vsd,
  rna_ajap,
  RNA_ajap,
  ajap_cl,
  ajap_cl_avgs,
  aj_aenja,
  file = "outputs/rda/ajap_reanalysis.rda"
)
```

```{r save table_cl, echo  = FALSE, eval = FALSE}
write.table(
  ajap_cl,
  paste(
    "outputs/comparative/reanalysis/ajap/", # change path
    fcha(),
    "_aj_dev_cl.tsv",
    sep = ""
  ),
  row.names = F,
  quote = F,
  sep = "\t",
  col.names = F
)
```