---
title: "Ptychodera Cisreg Development: Gene Age Analysis"
author: "Alberto Perez-Posada"
date: "3/31/2023"
output:
  html_document: default
  pdf_document: default
editor_options: 
  markdown: 
    wrap: sentence
---

```{r setup, include=FALSE}

dir <- '/home/ska/aperpos/projects/ptychodera_cisreg_development/'
fcha <- function(){ gsub("-","",Sys.Date()) }

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = dir)
options(scipen=999)
Sys.setenv(VROOM_CONNECTION_SIZE=5000072)

```


## About

In this markdown we will explore the gene age of the gene expression programme during the development of Ptychodera flava.

For this we will do:

 - Gene Age Enrichment Analysis, using the clusters of stage-specific genes
 - Transcriptome Age Index Analysis, using the trasncriptome counts of all genes of Ptychodera

For the two analyses, we will need a gene age lookup table for all the genes of Ptychodera. This was generated using the OMA software and the Count Software (see corresponding markdown).

## Load libraries

```{r libraries, message=FALSE, warning=FALSE}
library(RColorBrewer)
library(myTAI)
```

## Load functions

```{r load_functions, warning = FALSE, message = FALSE}
source("code/r_code/r_functions/sourcefolder.R")

sourceFolder(
  "code/r_code/r_functions",
  recursive = TRUE
  )
```

## Data Load

We will load the output of the DESEq2 markdown that contains the count tables and the stage-specific cluster membership lookup table.

```{r load data}
load("outputs/rda/stage_specific_clusters.rda")
```

We will also load the lookup table of ptychodera / OMA orthologous group that we will use to associate genes to a given age.

```{r load pfla_oma}
pfla_oma <-
  read.delim2(
    "outputs/comparative/20240404_oma/oma_omaid_gfam.tsv",
    # col.names = c("oma","id"),
    header = TRUE
  )
pfla_oma <- pfla_oma[grep("TCONS",pfla_oma$omaid),]
rownames(pfla_oma) <- NULL
pfla_oma$omaid <- gsub("\\.p[0-9]+","",pfla_oma$omaid)
head(pfla_oma)
```

We will also load the output of Count Software that provides presence/absence of each orthogroup in each species used for the comparative analysis AS WELL AS the Dolloparsimony-inferred gene age for each orthogroup.

```{r load dollo}
oma_dollo <-
  read.delim2(
    #"outputs/comparative/count/dollo.tsv",
    "outputs/comparative/dollo_gene_families/oma_dollo.tsv",
    sep = "\t",
    skip = 1, # commented lines
    header = T
  )

head(oma_dollo[,1:10])
```

And finally a small lookup table that we created by visually inspecting the tree of species and inferring what nodes in the tree correspond to our gene ages of interest.

```{r load geneage lookup}
ages_table <- 
  read.table(
    #"outputs/comparative/count/oma_geneage_table.tsv",
    "outputs/comparative/dollo_gene_families/dollo_nodes_names.tsv",
    header = TRUE
    )

nodes_of_interest_pfla <-
  c(
    "Pfla_Specific","hemichordata","ambulacraria",
    "deuterostomia","bilateria","eumetazoa","holozoa",
    "opisthokonta"
  )

ages_table <- ages_table[
  ages_table$name %in% nodes_of_interest_pfla,
]
colnames(ages_table) <- c("Node","Name")
head(ages_table)
```

## Gene Age table

We will first parse the Dollo table using our gene age lookup table and retrieving a lookup table of orthogroup/gene age. This custom wrapper can retrieve the gene age in an ascending or descending order on demand.

```{r parse geneage table}
gfam_age <- 
  parse_geneage_table(
    dollo_table = oma_dollo,
    ages_table = ages_table,
    reverse_order = TRUE
  )

head(gfam_age)
```

This we can merge with our geneid/orthogroup lookup table to finally retrieve the gene age of each ptychodera gene. We will manually add all ptychodera genes that were no present in that table as "Ptychodera-specific".

```{r get pfla_age}
pfla_age <-
  merge(
    pfla_oma,
    gfam_age,
    by.x = 2,
    by.y = 1,
    all.x = TRUE
  )[,c(2,3,4)]

pfla_age <-
  rbind(
    pfla_age,
    data.frame(
      omaid = allgenes[!(allgenes %in% pfla_age[,1])],
      age_num = 8,
      age = "8_Pfla"
    )
  )

colnames(pfla_age) <- c("id","age_num","age")

head(pfla_age)
```

## Gene Age Enrichment

A quick look at the gene age composition of all Ptychodera genes vs those expressed during development:

```{r piecharts}
par(mfrow = c(1,2))

pie(
  table(pfla_age$age),
  col = brewer.pal(8,"Spectral")
)

pie(
  table(pfla_age$age[pfla_age$id %in% rownames(pfla_rna_dev)]),
  col = brewer.pal(8,"Spectral")
)

par(mfrow = c(1,1))
```

And finally a custom wrapper function that performs Fisher's exact test for each gene age and each cluster of stage-specific expression during development.

```{r gene_age_enrichment}
pfla_age_enrichment <- gene_age_enrichment(
  x_modules = data.frame(
    id = rownames(pfla_rna_dev),
    module = pfla_rna_dev$cID),
  x_age = pfla_age[,c(1,3)],
  phylostrata = FALSE,
  fisher_pval = 0.01
  ) 
```

The visualisation of the heatmap:

```{r Gene Age Heatmap, fig.width = 6.5, fig.height = 6.5}
draw(pf_avg_hm+pfla_age_enrichment$heatmap)
```
And here using a chi-squared and post-hoc chi-squared test:


```{r, fig.width = 3, fig.height = 5}
cl_age_tbl = pfla_age_enrichment$AgeperModule
cl_age_chisq = chisq_and_posthoc(tbl = cl_age_tbl)
cl_age_chisq_hm = make_chisq_heatmap(cl_age_chisq)

mat <- cl_age_chisq_hm$residuals_matrix
mat <- t(mat)
mat <- mat[,order(colnames(mat))]

pvmat <- cl_age_chisq_hm$pvalues_matrix
pvmat <- t(pvmat)
pvmat <- pvmat[,order(colnames(pvmat))]

fisher_pval <- .05

cl_age_chisq_hm <- Heatmap(
  name = "residuals",
  mat,
  cluster_rows = FALSE,
  clustering_method_rows = "ward.D2",
  cluster_columns = FALSE,
  # row_km = 3,
  column_names_side = "top",
  row_names_side = "left",
  col = make_enr_pal(values = mat),
  cell_fun = function(j,i,x,y,width,height,fill){
    if(pvmat[i,j] < fisher_pval)
      grid.text(
        "*", x, y, gp = gpar(fontsize = 15)
      ) # add asterisk if p < .05
  }
)

draw(cl_age_chisq_hm)
```

## Transcriptome Age Index

We will create a data frame of gene/age in a MyTAI-compliant format.

```{r pfla_age_tai}
pfla_tai_age <- 
  data.frame(
    Phylostratum = as.numeric(pfla_age$age_num),
    GeneID = pfla_age$id
  )

head(pfla_tai_age)
```

We do the same with the table of counts during development.

```{r pfla_counts_tai}
pfla_tai_expr <- 
  data.frame(
    GeneID = rownames(pfla_rna_counts),
    pfla_rna_counts
    )

colnames(pfla_tai_expr) <- 
  gsub("^X","",colnames(pfla_tai_expr))

head(pfla_tai_expr[,1:5])
```

We create a PhyloExpressionSet object from the MyTAI package with both datasets we just created

```{r phyloexpressionset}
pfla.PhyloExpressionSet <- 
  MatchMap(pfla_tai_age,pfla_tai_expr)

is.ExpressionSet(pfla.PhyloExpressionSet) 
```

And finally the flatline test and the visualisation of the Transcriptome Age Index:

```{r mytai flatlinetest}
PlotSignature( 
  ExpressionSet = pfla.PhyloExpressionSet,
  measure       = "TAI", 
  TestStatistic = "FlatLineTest",
  modules       = list(early = 1:7, mid = 8:17, late = 18:31),
  xlab          = "Ontogeny", 
  ylab          = "TAI" 
  )

pdf("graphics/TAI.pdf", width = 6, height = 4)
PlotSignature( 
  ExpressionSet = pfla.PhyloExpressionSet,
  measure       = "TAI", 
  TestStatistic = "FlatLineTest",
  modules       = list(early = 1:7, mid = 8:17, late = 18:31),
  xlab          = "Ontogeny", 
  ylab          = "TAI" 
  )
dev.off()
```

## Saving the data

We will save the data in case we need it later:

```{r save}
save(
  pfla_age,
  pfla_oma,
  pfla_age_enrichment,
  pfla.PhyloExpressionSet,
  file = "outputs/rda/geneage.rda"
)
```
