---
title: "Ptychodera Cisreg Development: Data Load and Normalisation"
author: "Alberto Perez-Posada"
date: "3/24/2023"
output:
  html_document: default
  pdf_document: default
editor_options: 
  markdown: 
    wrap: sentence
---

```{r setup, include=FALSE}

dir <- '/home/ska/aperpos/projects/ptychodera_cisreg_development/'
fcha <- function(){ gsub("-","",Sys.Date()) }

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = dir)
options(scipen=999)
Sys.setenv(VROOM_CONNECTION_SIZE=5000072)

```


## About

In this markdown we will load the mapped RNA-seq counts of Ptychodera and normalise/transform the data, Using edgeR and RUV-Seq,  for proper downstream analysis.

## Load libraries

```{r libraries, message=FALSE, warning=FALSE}
library(DESeq2)
library(tximport)
library(RUVSeq)
library(rhdf5)
library(apeglm)
library(gplots)
library(pheatmap)
library(scales)
```

## Load functions

```{r load_functions, message=FALSE, warning=FALSE}
#from @Rich Scriven https://stackoverflow.com/questions/26045478/source-r-scripts-in-different-folders
sourceFolder <- function(folder, recursive = FALSE, ...) 
{ 
    files <- list.files(folder, pattern = "[.][rR]$", 
                        full.names = TRUE, recursive = recursive)
    if (!length(files))
        stop(simpleError(sprintf('No R files in folder "%s"', folder)))
    src <- invisible(lapply(files, source, ...))
    message(sprintf('%s files sourced from folder "%s"', length(src), folder))
}

sourceFolder(
  "code/r_code/r_functions",
  recursive = TRUE
  )

sourceFolder(
  "code/r_code/r_general",
  recursive = TRUE
  )


```


## Data Load

Preliminary DESeq2 design to retrieve the non-normalized count matrix.

```{r load table}

samplesdir <-
  "data/RNA/kallisto/abundances/"

sampleTable <-
  read.table(
    "data/RNA/kallisto/abundances/sample_table.tsv",
    header = F
  )[, c(1, 1, 2, 3,4)]

colnames(sampleTable) <-
  c("name", "dir", "condition", "batch0","batch")

sampleTable
```



```{r load files}
files <-
  setNames(
    object = list.files(
      samplesdir,
      "h5",
      recursive = T,
      full.names = T
      ),
    list.dirs(
      samplesdir,
      full.names = F
      )[-1]
  )

txi.kallisto <-
  tximport(
    files,
    type = "kallisto",
    txOut = TRUE
    )

dds_Txi <-
  DESeqDataSetFromTximport(
    txi.kallisto,
    colData = sampleTable,
    design = ~ batch + condition
    )

dds_Txi$condition <-
  relevel(dds_Txi$condition, ref = c("00_Uf"))

```
## Normalisation using edgeR and RUV-Seq

### Data preparation from the preliminary DESeq2 data

We will extract the raw counts from this newly-generated DESeq2 object. We will apply a filter to keep only the genes that have more than five counts in at least two columns (in our case, developmental stages)

```{r extract_counts}

counts_nonorm <- counts(dds_Txi[,c(1:31)],normalized=F)

allgenes <- rownames(counts_nonorm)
  
filter <- 
  apply(
    counts_nonorm, 1,
    function(x) length(x[x>5])>=2
    )

counts_nonorm_filt <- counts_nonorm[filter,]

condition_x <- as.factor(substr(sampleTable$condition,1,5))
```

We define an edgeR expressionset:

```{r edgeR exprset}
set <- newSeqExpressionSet(as.matrix(counts_nonorm_filt),
                           phenoData = data.frame(condition_x, row.names=colnames(counts_nonorm_filt)))

set
```

And the structure of the data in this expressionset:

```{r plot counts pca}
par(mfrow=c(1,2))
plotRLE(
  set,
  outline=FALSE,
  ylim=c(-4, 4),
  col=dev_palette[condition_x],
  cex.lab=0.5,
  main="Pre-UpperQuartile Normalization",
  xaxt="n",
  ylab="RNA-Seq counts (scaled)"
  )

axis(side = 1, at=1:length(colnames(set)), labels = FALSE)
text(x = 1:length(colnames(set)),
     ## Move labels to just below bottom of chart.
     y = par("usr")[3] - 0.45,
     ## Use names from the data list.
     labels = colnames(set),
     ## Change the clipping region.
     xpd = NA,
     ## Rotate the labels by 35 degrees.
     srt = 90,
     ## Adjust the labels to almost 100% right-justified.
     adj = 0.965,
     ## Increase label size.
     cex = 0.8)

plotPCA(
  set,
  col=dev_palette[condition_x],
  cex=0.75,
  main="Pre-UpperQuartile Normalization"
  )
set0 = set # for plotting the original
```

As we can see the count distribution varies widely across samples. We will address this doing upper quantile normalisation.

### Upper Quantile Normalisation

```{r quantilenorm}
set <- betweenLaneNormalization(set, which="upper")
```

And after Upper Quantile Normlisation:

```{r plot quantilenorm }
par(mfrow = c(1,2) )

plotRLE(
  set,
  outline=FALSE,
  ylim=c(-4, 4),
  col=dev_palette[condition_x],
  main="Post-UpperQuartile Normalization",
  xaxt="n",
  ylab="RNA-Seq counts (scaled)"
  )

axis(side = 1, at=1:length(colnames(set)), labels = FALSE)

text(x = 1:length(colnames(set)),
     ## Move labels to just below bottom of chart.
     y = par("usr")[3] - 0.45,
     ## Use names from the data list.
     labels = colnames(set),
     ## Change the clipping region.
     xpd = NA,
     ## Rotate the labels by 35 degrees.
     srt = 90,
     ## Adjust the labels to almost 100% right-justified.
     adj = 0.965,
     ## Increase label size.
     cex = 0.8)

plotPCA(
  set,
  col=dev_palette[condition_x],
  cex=0.75,main="Post-UpperQuartile Normalization"
  )
```

Things are looking much better now.

### Remove Unwanted Variation with RUV-Seq

First, edgeR to identify in-silico empirical genes that do not change, using linear modeling.

We will create a design matrix using `model.matrix` to let it know the conditions that are the same (==replicates of the same developmental stages). Then we create a `DGEList` object and calculate normalised factors and estimate dispersion (using `calcNormFactors` and `estimateGLMCommonDisp` and `estimateGLMTagwiseDisp`).

We will fit the data to the design matrix.

```{r glmfit edgeR}
design <- model.matrix(~condition_x, data=pData(set))

y <- DGEList(counts=counts(set), group=condition_x)

y <- calcNormFactors(y, method="upperquartile")
y <- estimateGLMCommonDisp(y, design)
y <- estimateGLMTagwiseDisp(y, design)

fit <- glmFit(y, design)
lrt <- glmLRT(fit, coef=2)
top <- topTags(lrt, n=nrow(set))$table

top[1:10,]
```

From this, we can identify those genes who do NOT change across conditions.

```{r empirical_genes}
empirical <- 
  rownames(set)[
    which(
      !(rownames(set) %in% rownames(top)[1:10000])
      )
    ]

empirical[1:10]
```

We can pass this down to RUV-Seq as in-silico determined invariant genes. Based on this, RUV-Seq will transform the data to account for the unwanted variation as estimated by changes in the in-silico invariant genes.

```{r RUVseq}
set2 <- RUVg(set, empirical, k=1)

pData(set2)
```

A look at the normalised counts after removing unwanted variation.

```{r plot after RUV}
par(mfrow = c(1,2))
plotRLE(
  set2,
  outline=FALSE,
  ylim=c(-4, 4),
  col=dev_palette[condition_x],
  main="Unwanted Variation Removed by RUVSeq",
  xaxt="n",
  ylab="RNA-Seq counts (scaled)"
  )

axis(side = 1, at=1:length(colnames(set)), labels = FALSE)

text(x = 1:length(colnames(set2)),
     ## Move labels to just below bottom of chart.
     y = par("usr")[3] - 0.45,
     ## Use names from the data list.
     labels = colnames(set2),
     ## Change the clipping region.
     xpd = NA,
     ## Rotate the labels by 35 degrees.
     srt = 90,
     ## Adjust the labels to almost 100% right-justified.
     adj = 0.965,
     ## Increase label size.
     cex = 0.8)

plotPCA(
  set2,
  col=dev_palette[condition_x],
  cex=0.75,
  main="Unwanted Variation Removed by RUVSeq"
  )

```


We finally generate a plain gene expression table with counts:

```{r}
pfla_rna_counts <-
  counts(set2)
```

We will save all the pertinent data in a rda for further analysis using DESeq2:


```{r save data}
save(
  # final counts
  set2,
  pfla_rna_counts,
  # condition
  condition_x,
  # sample table
  sampleTable,
  # list of all genes
  allgenes,
  file = "outputs/rda/normalised_counts.rda"
)

```

```{r echo = FALSE}
pdf(
  file = "graphics/1B.pdf",
  height = 4.1,
  width = 4
)
plotPCA(
  set0,
  col=dev_palette[condition_x],
  cex=0.75,
  main="PCA Develomental\nTranscriptomes",
  bty = "n"
)
dev.off()
```


```{r echo = FALSE}
pdf("graphics/supp_01_1.pdf", wi = 6, height = 9)
par(mfrow = c(3,2))
plotRLE(
  set0,
  outline=FALSE,
  ylim=c(-4, 4),
  col=dev_palette[condition_x],
  main="Pre-upper quantile normalsation",
  xaxt="n",
  ylab="RNA-Seq counts (scaled)"
)

axis(side = 1, at=1:length(colnames(set)), labels = FALSE)

text(x = 1:length(colnames(set0)),
     ## Move labels to just below bottom of chart.
     y = par("usr")[3] - 0.45,
     ## Use names from the data list.
     labels = colnames(set0),
     ## Change the clipping region.
     xpd = NA,
     ## Rotate the labels by 35 degrees.
     srt = 90,
     ## Adjust the labels to almost 100% right-justified.
     adj = 0.965,
     ## Increase label size.
     cex = 0.8)

plotPCA(
  set0,
  col=dev_palette[condition_x],
  cex=0.75,
  main="Pre-upper quantile Normalisation (PCA)"
)

plotRLE(
  set,
  outline=FALSE,
  ylim=c(-4, 4),
  col=dev_palette[condition_x],
  main="Post-UpperQuartile Normalization",
  xaxt="n",
  ylab="RNA-Seq counts (scaled)"
)

axis(side = 1, at=1:length(colnames(set)), labels = FALSE)

text(x = 1:length(colnames(set)),
     ## Move labels to just below bottom of chart.
     y = par("usr")[3] - 0.45,
     ## Use names from the data list.
     labels = colnames(set),
     ## Change the clipping region.
     xpd = NA,
     ## Rotate the labels by 35 degrees.
     srt = 90,
     ## Adjust the labels to almost 100% right-justified.
     adj = 0.965,
     ## Increase label size.
     cex = 0.8)

plotPCA(
  set,
  col=dev_palette[condition_x],
  cex=0.75,main="Post-UpperQuartile Normalisation (PCA)"
)

plotRLE(
  set2,
  outline=FALSE,
  ylim=c(-4, 4),
  col=dev_palette[condition_x],
  main="Unwanted Variation Removed by RUVSeq",
  xaxt="n",
  ylab="RNA-Seq counts (scaled)"
)

axis(side = 1, at=1:length(colnames(set)), labels = FALSE)

text(x = 1:length(colnames(set2)),
     ## Move labels to just below bottom of chart.
     y = par("usr")[3] - 0.45,
     ## Use names from the data list.
     labels = colnames(set2),
     ## Change the clipping region.
     xpd = NA,
     ## Rotate the labels by 35 degrees.
     srt = 90,
     ## Adjust the labels to almost 100% right-justified.
     adj = 0.965,
     ## Increase label size.
     cex = 0.8)

plotPCA(
  set2,
  col=dev_palette[condition_x],
  cex=0.75,
  main="Unwanted Variation Removed by RUVSeq (PCA)"
)
par(mfrow = c(1,1))
dev.off()
```