---
title: "Ptychodera Cisreg Development: ATAC Open Chromatin Regions analysis"
author: "Alberto Perez-Posada"
date: "4/4/2023"
output: html_document
---

```{r setup, include=FALSE}
dir <- '/home/ska/aperpos/projects/ptychodera_cisreg_development/'
fcha <- function(){ gsub("-","",Sys.Date()) }

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = dir)
options(scipen=999)
Sys.setenv(VROOM_CONNECTION_SIZE=5000072)

```


## About

In this markdown, we will do a descriptive analysis of the different regions of open chromatin that we found when mapping using bowtie2, macs2 and IDR.

This includes also associating the peaks to their closest gene to classify them according to their distance to genes, and see if they have similar or different properties among them.

Unless otherwise specified, I will use here the terms OCR, or cis-regulatory elements (CREs), or peaks, interchangeably.

## Load libraries

```{r libraries, message=FALSE, warning=FALSE}
library(DESeq2)
library(limma)
library(ComplexHeatmap)
library(circlize)
library(ggplot2)
library(ggpubr)
library(tidyverse)
library(stringr)
```

## Load functions

```{r load_functions, message = FALSE}
source("code/r_code/r_functions/sourcefolder.R")

sourceFolder(
  "code/r_code/r_functions",
  recursive = TRUE
  )

sourceFolder(
  "code/r_code/r_general",
  recursive = TRUE
  )
```

## Load the data

We will load the DESeq2 data for both the RNA-seq and the ATAC-seq.

```{r}
# load normalised counts, allgenes etc.
load("outputs/rda/normalised_counts.rda")
# load rna vsd
load("outputs/rda/deseq2.rda")
# load atac vsd
load("outputs/rda/atacseq_analysis.rda")
```

Here we will load the association of genes (specifically, the TSS of genes) and peaks that we did using `bedtools closestbed`.

```{r}
allpeaks_assoc_allgenes <-
  read.delim(
    "data/ATAC/closestbed/closestbed_allpeaks_tss.bed",
    header = F,
    stringsAsFactors = F
  )

colnames(allpeaks_assoc_allgenes) <-
  c(
    "chr", "start","end","id_peak",
    "ch_gene","st_gene","nd_gene","id_gene","V9","sense_gene",
    "distance"
  )
```

A quick look at the table:

```{r}
head(allpeaks_assoc_allgenes)
```

## Types of OCR peaks by relative position to genes/gene features 

We will remove the peaks that are further than 100 kilobases away for clarity purposes.

```{r}
peaks_genes_less100kb <- 
  allpeaks_assoc_allgenes[
    abs(allpeaks_assoc_allgenes$distance) < 100000,
    ]
par(mfrow = c(1,2))
plot(
  density(allpeaks_assoc_allgenes$distance),
  main = "Density distribution of distance of ATAC peaks to a gene"
  )
plot(
  density(peaks_genes_less100kb$distance),
  main = "Density distribution of distance of ATAC peaks to a gene"
  )
par(mfrow = c(1,1))
```

Here we can also estimate the n-percentiles of the peaks based on their distance to their closest gene.

```{r}
q_disttogene <- 
  quantile(
    abs(allpeaks_assoc_allgenes$distance),
    c(0,0.2,0.25,0.5,0.75,0.9)
    )
```

We can plot the cumulative density function of the peaks based on the distance to their closest gene.

```{r}
plot(
  ecdf(peaks_genes_less100kb$distance), #(seq(-99778,99959,by=500)),
  col="darkturquoise",
  main = "Cumulative density distribution of\ndistance of ATAC peaks to genes",
  xlab = "Distance to gene (bp)",
  ylab = "Cumulative density",
  xlim=c(-5,5)*10000
) #can also be seen in the cumul density distribution

abline(
  v=sort(unname(c(-1*q_disttogene,q_disttogene))),
  col=c(sequential_hcl(6,"GnBu"),rev(sequential_hcl(6,"GnBu"))),
  lwd=1.5
  )
```

We can also bin the genes based on these N-percentiles and count how many of each are in each bin of distance.

```{r}
# Define the quantile-based breakpoints
breaks <- q_disttogene

# Use cut() to split the distances into quantile-based categories
distance_categories <- cut(
  abs(allpeaks_assoc_allgenes$distance),
  breaks = c(breaks, Inf),
  labels = c("0-600bp", "600b-1.1kb", "1.1-6.2kb", "6.2-15.5kb", "15-29Kb", "beyond 29Kb"),
  include.lowest = TRUE
)

# Create a table of the distance categories and calculate their proportions
peaks_by_distance <- 
  table(distance_categories) / 
    length(allpeaks_assoc_allgenes$distance)

# Convert the table to a matrix and set the row names
peaks_by_distance <- as.matrix(peaks_by_distance)
rownames(peaks_by_distance) <- 
  c("0-600bp", "600-1.1Kb", "1.1-6.2Kb", 
    "6.2-15.5Kb", "15-29Kb", "beyond 29Kb")
```

Likewise, we did a parallell classification of peaks using the `AnnotatePeaks.pl` tool from the HOMER suite. Based on similar criteria, this tool is able to classify the peaks according to their relative position to the TSS of genes as well as if they fall within an exon, intron, etc feature.

```{r}
annot_stats_homer <- 
  read_tsv(
    file = "outputs/homer/annotatepeaks/all_peaks_annotatepeaks.tsv"
    )

head(annot_stats_homer)
```

We can plot the results of this binning here:

```{r}
par(mfrow = c(1,2))
barplot(
  peaks_by_distance,
  las=1,
  col=rev(sequential_hcl(6,"GnBu")),
  beside=F,
  xlim=c(0,11)
)
legend(
  2,0.6,
  rownames(peaks_by_distance),
  fill = rev(sequential_hcl(6,"GnBu")),
  cex=0.8
  )
barplot(
  matrix(
    annot_stats_homer$`Number of peaks`/
      sum(annot_stats_homer$`Number of peaks`)
    ),
  las=1,
  col=rev(sequential_hcl(5,"Red-Purple")),
  beside=F,
  xlim=c(0,11)
)
legend(
  2,0.6,
  annot_stats_homer$Annotation,
  fill = rev(sequential_hcl(5,"Red-Purple")),
  cex=0.8
  )
par(mfrow = c(1,1))
```

We will write in disk the bed files of the different types of peaks by category to perform a motif enrichment analysis with HOMER. See this below.

```{r, eval = FALSE}
# write peaks by category for homer
```

Lorem ipsum

## Analysis of genes by number of associated OCR/cRE

We will parse the gene/OCR association table to retrieve the genes based on their number of associated cis-regulatory elements. Genes that are not associated to any cRE are still added with `0` as their `Freq` value.

```{r}
numRE_gene <-merge(
  data.frame(
    id = allgenes
  ),
  data.frame(
    table(allpeaks_assoc_allgenes$id_gene)
    ),
  by.x = 1,
  by.y = 1,
  all.x = TRUE
)

numRE_gene$Freq[which(is.na(numRE_gene$Freq))] <- 0

head(numRE_gene)
```

For visualisation purposes, and ot prevent any perturbation to the original table, I copy this table and re-bin the genes as having 1, 2, 3-to-5, 6-to-10, and more-than-10 associated cREs.

```{r}
numRE_gene_viz <- numRE_gene

numRE_gene_viz$Freq <-
  ifelse(numRE_gene_viz$Freq > 10, "5",
  ifelse(numRE_gene_viz$Freq >= 6, "4",
  ifelse(numRE_gene_viz$Freq >= 3, "3",
  ifelse(numRE_gene_viz$Freq == 2, "2",
  ifelse(numRE_gene_viz$Freq == 1, "1",
  "0")))))

numRE_gene_viz <- 
  as.matrix(table(numRE_gene_viz$Freq))

rownames(numRE_gene_viz) <- 
  c("0","1","2","3-5","6-10",">10")

head(numRE_gene_viz)
```

And a barplot here:

```{r}
barplot(
  numRE_gene_viz/sum(numRE_gene_viz),
  las=1,
  col=rev(sequential_hcl(6,"GnBu")),
  beside=F,
  xlim=c(0,11),
  main = "Number of genes based on\ntheir number of associated OCRs"
)
legend(
  2,0.8,
  rev(rownames(numRE_gene_viz)),
  fill = sequential_hcl(6,"GnBu"),
  cex=0.8
)
```

## Number of RE/gene tests

Since genes controlling embryonic development tend to have a tightly-orchestrated regulation of their gene expression (at least in other organisms), one hypothesis is that this is also achieved in Ptychodera by being under the influence of a larger number of cRE compared to other types of genes.

For this, we annotated the Ptychodera genes as "trans-developmental" or "housekeeping" using an expanded definition from Marletaz et al., 2018, based on the types of Gene Ontologies they are associated with.

This table and the respective markdown can be found elsewhere in the directories.

```{r}
# load table of Trans-developmental / housekeeping
pfla_td_nohk <- 
  unique(
    read.table(
      "outputs/functional_annotation/TD_annotation/transdev_expanded_nohk.tsv",
      header=F,
      stringsAsFactors = F)[,1]
    )

pfla_hk_notd <- 
  unique(
    read.table(
      "outputs/functional_annotation/TD_annotation/housekeep_expanded_notd.tsv",
      header=F,
      stringsAsFactors=F)[,1]
    )

pfla_tdhk <- data.frame(
  id = allgenes,
  TDHK =  
    ifelse(allgenes %in% pfla_td_nohk, "td",
    ifelse(allgenes %in% pfla_hk_notd, "hk", 
    "none")
    )
  )
# make table of pfla_tdhk_numRE using merge by.x = 1, by.y = 1
pfla_tdhk_numRE <-
  merge(
    pfla_tdhk,
    numRE_gene,
    by.x = 1,
    by.y = 1,
    all = TRUE
  )
```

We can make a contingency table to check this using a fisherÂ´s Exact test. We will first check if there is a higher-than-chance proportion of genes with associated Re that are trans-dev.

```{r}
fisher.numRE.transdev <- 
  matrix(c(
    nrow(pfla_tdhk_numRE[(pfla_tdhk_numRE$Freq == 0 & pfla_tdhk_numRE$TDHK != "td"),]),
    nrow(pfla_tdhk_numRE[(pfla_tdhk_numRE$Freq == 0 & pfla_tdhk_numRE$TDHK == "td"),]),
    nrow(pfla_tdhk_numRE[(pfla_tdhk_numRE$Freq != 0 & pfla_tdhk_numRE$TDHK != "td"),]),
    nrow(pfla_tdhk_numRE[(pfla_tdhk_numRE$Freq != 0 & pfla_tdhk_numRE$TDHK == "td"),])
    ), nrow=2)

colnames(fisher.numRE.transdev) <-
  c("!RE","RE")

rownames(fisher.numRE.transdev) <-
  c("!td","td")

fisher.numRE.transdev

fisher.test(fisher.numRE.transdev)
```

The ratio of being trans-dev and having associated RE is significantly larger than no RE and being trans-dev.

We can go one step further and check if, of all the genes with associated REs, there is a higher chance to be trans-dev if a gene has more than one associated cRE.

```{r}
# % OF TRANSDEV ASSOC TO 1(ONE) RE VS ASSOC TO 2 OR MORE RE

fisher.1vsmoreRE.transdev <- 
  matrix(c(
    nrow(pfla_tdhk_numRE[(pfla_tdhk_numRE$Freq == 1 & pfla_tdhk_numRE$TDHK != "td"),]),
    nrow(pfla_tdhk_numRE[(pfla_tdhk_numRE$Freq == 1 & pfla_tdhk_numRE$TDHK == "td"),]),
    nrow(pfla_tdhk_numRE[(pfla_tdhk_numRE$Freq > 1 & pfla_tdhk_numRE$TDHK != "td"),]),
    nrow(pfla_tdhk_numRE[(pfla_tdhk_numRE$Freq > 1 & pfla_tdhk_numRE$TDHK == "td"),])
    ), nrow=2)

colnames(fisher.1vsmoreRE.transdev) = c("1RE","+RE")

rownames(fisher.1vsmoreRE.transdev) = c("!td","td")

fisher.1vsmoreRE.transdev

fisher.test(fisher.1vsmoreRE.transdev)
```

The ratio of being trans-dev and having more than one RE is significantly larger than being trans-dev and having just one assoc RE.

And the resulting barplots of the fraction of trans-dev genes:

```{r}
barplot(
  c(
    fisher.numRE.transdev[2, 1] / fisher.numRE.transdev[1, 1],
    fisher.numRE.transdev[2, 2] / fisher.numRE.transdev[1, 2],
    fisher.1vsmoreRE.transdev[2, 1] / fisher.1vsmoreRE.transdev[1, 1],
    fisher.1vsmoreRE.transdev[2, 2] / fisher.1vsmoreRE.transdev[1, 2]
  ),
  ylim = c(0, 0.05),
  names.arg = c("no assoc RE", "assoc RE","1 RE", "2 or more RE"),
  col = c("#ce8181", "#843c3c","#ce8181", "#843c3c"),
  main = "Fraction of trans-developmental genes\nassociated to cis-regulatory elements",
  ylab = "% genes",
)
lines(c(0.5, 2), c(0.04, 0.04))
text(1.2, 0.045, "***")
lines(c(3, 4.5), c(0.04, 0.04))
text(3.8, 0.045, "***")

```

### Comparison of Transdev/Housekeeping REs

Another way to see this is defining two populations of genes: those that are trans-dev and those that are housekeeping genes. We can compare if the distribution of associated cREs per gene come frmo the same population for both gene sets or if there are significant differences using a Wilcoxon test.

For this we will randomly subsample both gene sets.

```{r}
# Transdev-Housekeeping number of REs
set.seed(42)
list_numre_td_hk <-
  list(
    transdev = 
      sample(
        pfla_tdhk_numRE$Freq[pfla_tdhk_numRE$TDHK == "td"],
        400
      ),
    housekeeping = 
      sample(
        pfla_tdhk_numRE$Freq[pfla_tdhk_numRE$TDHK == "hk"],
        400
    )
  )

wilcox.test(
  pfla_tdhk_numRE$Freq[pfla_tdhk_numRE$TDHK == "td"],
  pfla_tdhk_numRE$Freq[pfla_tdhk_numRE$TDHK == "hk"]
  )
```

Besides a boxplot, this can also be observed using a cumulative density function of each gene set.

```{r}
par(mfrow = c(1,2))

#Boxplot
boxplot(
  list(
    housekeep = pfla_tdhk_numRE$Freq[pfla_tdhk_numRE$TDHK == "hk"],
    transdev = pfla_tdhk_numRE$Freq[pfla_tdhk_numRE$TDHK == "td"]
    ),
  notch=T,
  col=c(
    "#ecb8a6",
    "#e1798f"
  ),
  main="distribution of associated\nREs in TD and HK genes",
  cex.axis=0.9,
  outline=F,
  ylim=c(0,33)
)
stripchart(
  list(
    pfla_tdhk_numRE$Freq[pfla_tdhk_numRE$TDHK == "hk"],
    pfla_tdhk_numRE$Freq[pfla_tdhk_numRE$TDHK == "td"]
  ),
  method = "jitter",
  jitter=0.15,
  col = c(
    alpha("#f07847",0.5),
    alpha("#b11e31",0.5)
  ),
  vertical = TRUE,
  pch = 20,
  cex=0.7,
  add = TRUE
)

# Cumulative density function
plot(
  cbind(0:40, ecdf(list_numre_td_hk$housekeeping)(0:40)),
  type = "s",
  col = "#f07847",
  main = "Cumulative density",
  xlab = "Number of REs",
  ylab = "% of genes",
  cex.axis = 0.9,
  lwd = 1.8,
  ylim = c(0, 1)
)
lines(cbind(0:40, ecdf(list_numre_td_hk$transdev)(0:40)),
      type = "s",
      col = "#b11e31",
      lwd = 1.8)
par(mfrow=c(1,1))
```

With this, it is safe to say that we can discard the null hypothesis that there is no association of having associated cRE and being involved in the contrl of development, hinting at our initial thought that gene expression during embryonic development is regulated with cis regulatory elements.

## Differentially regulated atac peaks and their relationship to assoc genes

If that was the case, one would expect to see a concordance between the grade of accessibility of OCR and the expression level of their associated genes. For this we can check if there are changes or differences in chromatin accessibility between OCRs associated to lowly-, mid-, or highly-expressed genes.

We can bin genes by their level of expression in each of the developmental stages for which we have chromatin data, and visualise the accessibility of their associated cREs.

```{r}
# subset rna vsd
head(pfla_rna_all)

# subset atac vsd
head(pfla_atac)

allpeaks_assoc_diffgenes <- 
  allpeaks_assoc_allgenes[
    allpeaks_assoc_allgenes$id_gene %in% rownames(vsd_signif),
    c(4, 8)
    ]

atac_assoc_genes <-
  allpeaks_assoc_allgenes[
    allpeaks_assoc_diffgenes$id_peak %in% rownames(pfla_atac),
    c(4,8)
  ]

# define reusable variables
rna_cols <- c(3, 5, 6, 8)
atac_cols <- c(1:4)
quantiles <- lapply(pfla_rna_all[, rna_cols], quantile)

# define function for processing each stage of development
process_stage <- function(stage, allpeaks_assoc_diffgenes) {
  
  # grab the rna counts
  stage_rna <- setNames(pfla_rna_all[,rna_cols[stage]],rownames(pfla_rna_all))
  
  # bin by lo-, mid-, or hi- expressed genes
  stage_tier <- 
    setNames(
      cut(stage_rna, breaks = quantiles[[stage]][c(1,3:5)],
          labels = c("lo", "med", "hi"),include.lowest = TRUE),
      names(stage_rna)
    )
  
  # grab the atac counts
  stage_atac <- data.frame(
    peak_id = rownames(pfla_atac),
    counts = log(pfla_atac[,atac_cols[stage]]+1)
    )
  
  # associate to lo-, mid-, or hi- expressed genes using assoc genes table
  stage_counts <- 
    merge(
      stage_atac,
      merge(
        data.frame(id = names(stage_tier), tier = stage_tier, stage = stage),
        allpeaks_assoc_diffgenes,
        by.x = 1, by.y = 2
        ),
      by.x = 1,by.y = 4
    )
  stage_counts
}

# process each stage of development and combine results into a single data frame
atac_by_rna_tiers <- do.call(rbind, lapply(1:4, process_stage, allpeaks_assoc_diffgenes))

# set factor levels and plot the results
atac_by_rna_tiers$stage <- 
  factor(
    setNames(
      c("eb", "eg", "mg", "lg"),1:4)
    )[atac_by_rna_tiers$stage]
  #)

ggplot(data = atac_by_rna_tiers) +
  geom_boxplot(aes(x = stage, y = counts, fill = tier)) +
  scale_fill_manual(values = c(sequential_hcl(5, "YlGnBu")[4:2])) +
  theme_minimal() +
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

```


## Motif Similarity

To look for enriched motifs we need to use a tool like the HOMER suite. The thing is most of these tools are optimised or only contain data for model organisms like vaertebrates. Prior to any motif enrichment analysis, it would be a good idea to check if de novo motifs found in the genome of our species have any resemblance to known motifs in the databases of this tool.

For this we can run a de novo search using HOMER, which will also give us a similarity score to a known motif for any of the de novo motifs found. We set this tool to look for at least one hundred de novo motifs.

The reults can be loaded here and we can plot the density function of these similarity scores to see how our data is behaving.

```{r}

#' Also missing: de novo motif discovery and % similarity with already-known motifs

homer_identity <- read.delim2(
  "outputs/homer/denovo/homer_denovo_200_similarity.tsv",
  header=F
)

homer_identity <- homer_identity[homer_identity$V4!= "", ]
homer_identity$V4 <- as.numeric(homer_identity$V4)
plot(
  density(homer_identity$V4),
  main="Similarity of de novo and known motifs"
  )

# Normal curve
fun <- dnorm(homer_identity$V4, mean = mean(homer_identity$V4), sd = sd(homer_identity$V4))

hist(
  homer_identity$V4,
  prob=TRUE,
  # ylim = c(0, 5),
  breaks = 50,
  main="Similarity of de novo and known motifs"
  )
lines(density(homer_identity$V4), col = "lightseagreen", lwd = 2)
```

Having seen that a number of these de novo motifs are highly similar to wel known motifs, we can proceed to do motif enrichment analysis with the different subsets of OCRs we have defined.

```{r, eval = FALSE}
atac_category_homer <- 
  read.delim(
    "/home/ska/aperpos/projects/ptychodera_cisreg_development/outputs/homer/annotatepeaks/allpeaks_category.tsv",
    header=F, 
    col.names=c("peakID","category")
    )

# motif enrichment atac peaks by homer category
for (i in unique(atac_category_homer$category)[1:5]){
  tbl_to_write <- allpeaks_assoc_allgenes[allpeaks_assoc_allgenes$V4 %in% atac_category_homer$peakID[atac_category_homer$category== i],]
  write.table(
    tbl_to_write,
    paste0(
      "outputs/homer/annotatepeaks/peaks/",
      "annotatepeaks_category_",
      i,
      ".bed",
      ),
    sep="\t",
    dec=".",
    row.names=F,
    col.names=F,
    quote=F
  )
}

```

```{r, echo = FALSE}
atac_category_homer <- 
  read.delim(
    "/home/ska/aperpos/projects/ptychodera_cisreg_development/outputs/homer/annotatepeaks/allpeaks_category.tsv",
    header=F, 
    col.names=c("peakID","category")
    )
```

We can use a set of custom wrapper functions to load the concatenated output of findMotifsGenome.pl and visualise the percentage of target peaks with enriched motifs as well as the logqval of significance.

In the case of OCRs subdivided by category as annotated by the annotatePeaks.pl tool of HOMER:

```{r}
# Load table
pfla_motifs_peakcategory <- 
  read_tsv("outputs/homer/annotatepeaks/motif_enrichments/all_motifs_per_peak_category.tsv")

# tidyr wrapper to present a tidy table
pfla_motifs_peakcategory <- parse_homer_output_table(
  tsv_motifs = pfla_motifs_peakcategory,
  categ_regex1 = "..*category_",
  categ_regex2 = "",
  max_logqval = 10
  )

colnames(pfla_motifs_peakcategory)[1] <- "category"

pfla_motifs_percateg_perclass <- 
  aggregate_motifs_table(pfla_motifs_peakcategory)
```

And the actual plot:

```{r, fig.height = 16, fig.width = 8}
# Giant figure, suppl.
atac_motifs_homercat_minim <- 
  aggregate(
    pfla_motifs_peakcategory[,c(8,11,15)],
    by = list(
      motif = pfla_motifs_peakcategory$motif_minim,
      category = pfla_motifs_peakcategory$category
      ),
    FUN = max
    )

# provide ggplot a new ordering of the factor
atac_motifs_homercat_minim <- 
  atac_motifs_homercat_minim[
    order(
      atac_motifs_homercat_minim$category,
      atac_motifs_homercat_minim$motif
      ),
    ]

atac_motifs_homercat_minim$motif <-
  factor(
    atac_motifs_homercat_minim$motif,
    levels=unique(atac_motifs_homercat_minim$motif)
    )

ggplot(data = atac_motifs_homercat_minim, aes(x = category, y = motif, # esto ta der carajo
                                              color = `logqval`, size = as.factor(size_pct))) +
  geom_point() +
  scale_color_continuous_sequential(palette = "BluGrn")+
  # scale_color_gradient(low = "#f1c232", high = "#5b0d80") +
  theme_bw() +
  ylab("") +
  xlab("") +
  ggtitle("Motif enrichment analysis") +
  scale_y_discrete(limits=rev) +
  theme(text = element_text(size=10), legend.text = element_text(size=10), axis.text.x = element_text(angle = 90, vjust = 0.5)) +
  scale_size_manual(values = c(1, 2, 3, 4,5), labels = c("1-10", "10-20", "20-40", "40+")) +
  guides(size=guide_legend(title="% peaks with motif")
  )
```

A summarised figure with motif enrichments pooled at the class level (perhaps less informative than the other one).

```{r}
# Simpler version at the class level
ggplot(data = pfla_motifs_percateg_perclass, aes(x = category, y = class,
                                              color = `logqval`, size = as.factor(size_pct))) +
  geom_point() +
  scale_color_continuous_sequential(palette = "BluGrn")+
  # scale_color_gradient(low = "#f1c232", high = "#5b0d80") +
  theme_bw() +
  ylab("") +
  xlab("") +
  ggtitle("Motif enrichment analysis") +
  scale_y_discrete(limits=rev) +
  theme(text = element_text(size=10), legend.text = element_text(size=10), axis.text.x = element_text(angle = 90, vjust = 0.5)) +
  scale_size_manual(values = c(1, 2, 3, 4,5), labels = c("1-10", "10-20", "20-40", "40+")) +
  guides(size=guide_legend(title="% peaks with motif")
  )
```

## OCR analyisis on all vs diffreg peaks

```{r}
#' Missing here: the distance-to-tss/classification in all vs diffreg peaks
diffpeaks_assoc_allgenes <-
  allpeaks_assoc_allgenes[allpeaks_assoc_allgenes$id_peak %in% rownames(pfla_atac),]
  
# Use cut() to split the distances into quantile-based categories
diff_distance_categories <- cut(
  abs(diffpeaks_assoc_allgenes$distance),
  breaks = c(breaks, Inf),
  labels = c("0-600bp", "600b-1.1kb", "1.1-6.2kb", "6.2-15.5kb", "15-29Kb", "beyond 29Kb"),
  include.lowest = TRUE
)

# Create a table of the distance categories and calculate their proportions
diffpeaks_by_distance <- 
  table(diff_distance_categories) / 
    length(diffpeaks_assoc_allgenes$distance)

# Convert the table to a matrix and set the row names
diffpeaks_by_distance <- as.matrix(diffpeaks_by_distance)

all_vs_diffpeaks_by_distance <-
  matrix(
    c(peaks_by_distance,diffpeaks_by_distance),
    nrow = 6,
    ncol = 2
  )

rownames(all_vs_diffpeaks_by_distance) <- 
  c("0-600bp", "600-1.1Kb", "1.1-6.2Kb", 
    "6.2-15.5Kb", "15-29Kb", "beyond 29Kb")

colnames(all_vs_diffpeaks_by_distance) <- c("all", "diff")


diff_atac_category_homer <-
  atac_category_homer[
    atac_category_homer$peakID %in% rownames(pfla_atac),
  ]

all_vs_diff_atac_category_homer <-
  matrix(
    c(
      table(atac_category_homer$category)/nrow(atac_category_homer),
      table(diff_atac_category_homer$category)/nrow(diff_atac_category_homer)
    ),
    ncol = 2,
    nrow = 5
  )

rownames(all_vs_diff_atac_category_homer) <-
  c(
    "Exon",
    "Intergenic",
    "Intron",
    "Promoter/TSS",
    "TTS"
  )
colnames(all_vs_diff_atac_category_homer) <- c("all", "diff")
```

And the barplots:

```{r}
par(mfrow = c(1,2))
barplot(
  all_vs_diffpeaks_by_distance,
  las=2,
  col=rev(sequential_hcl(6,"GnBu")),
  beside=F,
  xlim=c(0,11)
)
legend(
  3,0.6,
  rownames(all_vs_diffpeaks_by_distance),
  fill = rev(sequential_hcl(6,"GnBu")),
  cex=0.8
  )


barplot(
  all_vs_diff_atac_category_homer,
  las=2,
  col=rev(sequential_hcl(5,"RdPu")),
  beside=F,
  xlim=c(0,11)
)
legend(
  3,0.6,
  rownames(all_vs_diff_atac_category_homer),
  fill = rev(sequential_hcl(5,"RdPu")),
  cex=0.8
  )
par(mfrow = c(1,1))
```
## Motif Enrichment Analysis of Differentially accessible peaks

We will proceed as above only using the results of the HOMER motif enrichment analysis for all the differentially accessible peaks, classified by cluster of differential accessibility throughout gastrulation.

## Motif Enrichment Analysis

We can use a set of custom wrapper functions to load the concatenated output of findMotifsGenome.pl and visualise the percentage of target peaks with enriched motifs as well as the logqval of significance. In the case of stage-specific clusters (this is also present in the differential accessibility markdown):

```{r}
# Load table
pfla_motifs_clusters <- 
  read_tsv("outputs/homer/atac_clusters/motif_enrichments/all_motifs_per_atac_cluster.tsv")

# tidyr wrapper to present a tidy table
pfla_motifs_clusters <- parse_homer_output_table(
  tsv_motifs = pfla_motifs_clusters,
  categ_regex1 = "..*category_",
  categ_regex2 = "",
  max_logqval = 10
  )

colnames(pfla_motifs_clusters)[1] <- "category"

pfla_motifs_clusters$category <- gsub(".*_cluster_", "", pfla_motifs_clusters$category)

atac_newcl <- data.frame(
  kmeans=c(5,2,1,6,3,4),
  newkl=c("01_EB","02_EB_EG","03_EG_MLG","04_EMG_LG","05_MG_LG","06_LG")
  )

pfla_motifs_clusters$category <- translate_ids(x = pfla_motifs_clusters$category, dict = atac_newcl)

pfla_motifs_percateg_perclass <- 
  aggregate_motifs_table(pfla_motifs_clusters)
```

And the actual plot:

```{r, fig.height = 6, fig.width = 12}
# Giant figure, suppl.
pfla_motif_clusters_minim <- 
  aggregate(
    pfla_motifs_clusters[,c(8,11,15)],
    by = list(
      motif = pfla_motifs_clusters$motif_minim,
      category = pfla_motifs_clusters$category
      ),
    FUN = max
    )

# provide ggplot a new ordering of the factor
pfla_motif_clusters_minim <- 
  pfla_motif_clusters_minim[
    order(
      pfla_motif_clusters_minim$category,
      pfla_motif_clusters_minim$motif
      ),
    ]

pfla_motif_clusters_minim$motif <-
  factor(
    pfla_motif_clusters_minim$motif,
    levels=unique(pfla_motif_clusters_minim$motif)
    )

ggplot(data = pfla_motif_clusters_minim, aes(x = motif, y = category, # esto ta der carajo
                                              color = `logqval`, size = as.factor(size_pct))) +
  geom_point() +
  scale_color_continuous_sequential(palette = "SunsetDark")+
  # scale_color_gradient(low = "#f1c232", high = "#5b0d80") +
  theme_bw() +
  ylab("") +
  xlab("") +
  ggtitle("Motif enrichment analysis") +
  scale_y_discrete(limits=rev) +
  theme(text = element_text(size=10), legend.text = element_text(size=10), axis.text.x = element_text(angle = 90, vjust = 0.5)) +
  scale_size_manual(values = c(1, 2, 3, 4,5), labels = c("1-10", "10-20", "20-40", "40+")) +
  guides(size=guide_legend(title="% peaks with motif")
  )
```

A summarised figure with motif enrichments pooled at the class level (perhaps less informative than the other one).

```{r}
pfla_motif_clusters_minim_perclass <- 
  aggregate_motifs_table(pfla_motifs_clusters)

pfla_motif_clusters_minim_perclass <-
  pfla_motif_clusters_minim_perclass[
    order(
      pfla_motif_clusters_minim_perclass$category,
      pfla_motif_clusters_minim_perclass$class
    ),
  ]

pfla_motif_clusters_minim_perclass$class <-
  factor(
    pfla_motif_clusters_minim_perclass$class,
    levels = unique(pfla_motif_clusters_minim_perclass$class)
    )

# Simpler version at the class level
ggplot(data = pfla_motif_clusters_minim_perclass, aes(x = class, y = category,
                                                color = `logqval`, size = as.factor(size_pct))) +
  geom_point() +
  scale_color_continuous_sequential(palette = "SunsetDark")+
  # scale_color_gradient(low = "#f1c232", high = "#5b0d80") +
  theme_bw() +
  ylab("") +
  xlab("") +
  ggtitle("Motif enrichment analysis") +
  scale_y_discrete(limits=rev) +
  theme(text = element_text(size=10), legend.text = element_text(size=10), axis.text.x = element_text(angle = 90, vjust = 0.5)) +
  scale_size_manual(values = c(1, 2, 3, 4,5), labels = c("1-10", "10-20", "20-40", "40+")) +
  guides(size=guide_legend(title="% peaks with motif")
  )

#' Also missing: homer motif analysis of differentially regulated proximal vs distal, all diffreg peaks as background
```
