---
title: "Ptychodera Cisreg Development: ATAC-seq Analysis"
author: "Alberto Perez-Posada"
date: "3/31/2023"
output:
  html_document: default
  pdf_document: default
editor_options: 
  markdown: 
    wrap: sentence
---

```{r setup, include=FALSE}

dir <- '/home/ska/aperpos/projects/ptychodera_cisreg_development/'
fcha <- function(){ gsub("-","",Sys.Date()) }

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = dir)
options(scipen=999)
Sys.setenv(VROOM_CONNECTION_SIZE=5000072)

```

## About

In this markdown, we analysed the counts of the Open Chromatin Regions (OCR, or 'peaks'), of our ATAC-Seq mapping. For this we will also use DESeq2.

## Load libraries

```{r libraries, message=FALSE, warning=FALSE}
library(DESeq2)
library(limma)
library(ComplexHeatmap)
library(circlize)
library(ggplot2)
library(tidyverse)
library(stringr)
```

## Load functions

```{r load_functions, message = FALSE}
source("code/r_code/r_functions/sourcefolder.R")

sourceFolder(
  "code/r_code/r_functions",
  recursive = TRUE
  )

sourceFolder(
  "code/r_code/r_general",
  recursive = TRUE
  )
```

## Load the data

The tsvs of the counts are all deposited in `data/ATAC/mapped/counts`.

We will create a table with all the sample information just like we did when analysing the RNA-seq data.

```{r}
sample_dir <- "data/ATAC/mapped/counts/"

sample_table <- 
  data.frame(
    sampleNames = 
      c(paste("dev_ATAC_pfla0", c(1:4,7,8,5,6), sep = "")), 
    fileName = 
      dir(path = sample_dir, pattern = "\\.counts"), 
    condition = 
      factor(c(c("EB","EB","EG","EG","LG","LG","MG","MG"))),
    batch = 
      factor(c(1,2,1,2,1,2,1,2))
  )
```

We will use the `DESeqDataSetFromHTSeqCount` function to load the counts of the tsv files. We will use the `~batch+condition` formula for the design matrix, and relevel the condition for `"EB"` since this is the first developmental stage.

```{r}
dds_atac <- 
  DESeqDataSetFromHTSeqCount(
    sampleTable = sample_table,
    directory = sample_dir,
    design = ~batch+condition
    )

dds_atac$condition <- 
  relevel(dds_atac$condition, ref = c("EB"))
```

Much like before since this is also time-series data, we will also use the Likelihood-Ratio test to determine the OCRs who showcase differential accessibility throughout these developmental stages.

```{r}
dds_atac <- 
  DESeq(
  dds_atac,
  test = "LRT",
  reduced = ~batch
  )
```

The resulting output here:

```{r}
res_atac <- results(dds_atac)

round(data.frame(res_atac)[1:10,],3)
```

We will retrieve the OCRs which show a significant change in chromatin accessibility during development.

```{r}
cutoff=0.05

res_signif <- 
  subset(res_atac, padj < cutoff )

res_signif <- 
  subset(res_signif, abs(log2FoldChange) > 1 )

res_signif <- 
  res_signif[ order(res_signif$padj), ]

head(sapply(res_signif,round,digits=3))
```

We extract the significant counts from the DESeq2 object using `assay()` and subsetting for those in the list of significant OCRs.

```{r}
counts_allpeaks <- 
  assay(
    dds_atac[,c(1:4,7,8,5,6)],
    normalized = T
    )

counts_signif <- 
  counts_allpeaks[ rownames(counts_allpeaks) %in% rownames(res_signif), ]
```

A PCA analysis of the eight samples:

```{r}
pi <- prcomp(t(counts_signif))

plot(
  pi$x,
  pch=16,
  col=alpha("black",0.5)
  )

text(
  pi$x[,c(1,2)],
  labels = 
    paste0(
      levels(sample_table$condition)[
        match(colnames(counts_signif),sample_table$sampleNames)
        ],
      as.character(sample_table$batch)[
        match(colnames(counts_signif),sample_table$sampleNames)
        ]
      ),
  pos=3,
  col="black",
  cex=0.6
  )
```

And Finally, we average by replicates and get the final set of significant OCR counts:

```{r}
pfla_atac <- 
  data.frame(
    EB=rowMeans(counts_signif[,1:2]),
    EG=rowMeans(counts_signif[,3:4]),
    MG=rowMeans(counts_signif[,5:6]),
    LG=rowMeans(counts_signif[,7:8])
  )
```

We subsample this dataset to visualise in a heatmap.

```{r}
pfla_atac_viz <-
  t(scale(t(
    pfla_atac[
      sample(nrow(pfla_atac), 500),
    ] 
    )))
```

We create the heatmap with the package `ComplexHeatmap`:

```{r}
pfla_atac_hm <-
  Heatmap(
    name = "chrom accs",
    pfla_atac_viz,
    show_row_names = FALSE,
    column_names_side="top",
    cluster_rows = TRUE,
    cluster_columns = FALSE,
    clustering_method_rows = "ward.D2",
    col=colorRampPalette(c("#337ab7","white","#f05757"))(10),
    )
```

And the actual visualisation:

```{r, fig.width = 3.5, fig.height = 5}
draw(pfla_atac_hm)
```

## K-means clustering

We can see there is around six big groups of differential accessibility. Using kmeans clustering we can retrieve these peaks classified in such clusters of differential accessibility.

```{r}
nclust <- 6

set.seed(1234)

k2 <- 
  kmeans(
    t(scale(t(pfla_atac))), 
    centers = nclust, 
    nstart = 25
  ) 
```

We add this info to the set of significant peaks:

```{r}
pfla_atac_kmeans <- 
  data.frame(
    t(scale(t(pfla_atac))),
    kmeans=k2$cluster
    )
```

We resssign a new cluster value (while keeping the structure) using match and a temporary little table.

```{r}
atac_newcl <- data.frame(
  kmeans=c(5,2,1,6,3,4),
  newkl=c("01_EB","02_EB_EG","03_EG_MLG","04_EMG_LG","05_MG_LG","06_LG")
  )

pfla_atac_kmeans$kmeans <-
  translate_ids(x = pfla_atac_kmeans$kmeans, dict = atac_newcl)
```

Again, we subset to visualise, create a Heatmap Object, and get the heatmap figure.

```{r, message = FALSE, warning = FALSE}
test <- 
  pfla_atac_kmeans[
    sample(nrow(pfla_atac_kmeans), 500),
  ]

pfla_atac_hm <- Heatmap(
  test[,1:4],
  name = "chrom accs",
  row_split=test$kmeans,
  show_row_names = F,
  row_title_side = "left",
  cluster_rows=F,
  cluster_row_slices=T,
  # row_dend_reorder = T,
  cluster_columns=F,
  row_title=NULL,
  column_names_side="top",
  use_raster = T,
  col=colorRampPalette(c("#337ab7","white","#f05757"))(10)
)
```

And the heatmap:

```{r, fig.width = 3.5, fig.height = 5}
draw(pfla_atac_hm)
```

## Writing bed files for motif enrichment

We write the bed files of the clusters to later do motif enrichment:

```{r eval = FALSE}
for (i in 1:nclust) {
  assign(paste("pfla_atac_cluster", i, sep = "."), pfla_atac_kmeans[which(pfla_atac_kmeans$kmeans ==i), ])
  
  numbercluster <- assign(paste("cluster", i, sep = "."), pfla_atac_kmeans[pfla_atac_kmeans$kmeans ==i, ])
  
  write.table(
    rownames(numbercluster),
    paste0(
      "outputs/atac/clusters_development/",
      "pfla_ATAC_cluster_",i,".txt"
      ),
    sep="\t",
    dec=".",
    row.names=F,col.names=F,quote=F
    )
}
```


## Motif Enrichment Analysis

We can use a set of custom wrapper functions to load the concatenated output of findMotifsGenome.pl and visualise the percentage of target peaks with enriched motifs as well as the logqval of significance. In the case of stage-specific clusters:

```{r}
# Load table
pfla_motifs_clusters <- 
  read_tsv("outputs/homer/atac_clusters/motif_enrichments/all_motifs_per_atac_cluster.tsv")

# tidyr wrapper to present a tidy table
pfla_motifs_clusters <- parse_homer_output_table(
  tsv_motifs = pfla_motifs_clusters,
  categ_regex1 = "..*category_",
  categ_regex2 = "",
  max_logqval = 10
  )

colnames(pfla_motifs_clusters)[1] <- "category"

pfla_motifs_clusters$category <- gsub(".*_cluster_", "", pfla_motifs_clusters$category)

atac_newcl <- data.frame(
  kmeans=c(5,2,1,6,3,4),
  newkl=c("01_EB","02_EB_EG","03_EG_MLG","04_EMG_LG","05_MG_LG","06_LG")
  )

pfla_motifs_clusters$category <- translate_ids(x = pfla_motifs_clusters$category, dict = atac_newcl)

pfla_motifs_percateg_perclass <- 
  aggregate_motifs_table(pfla_motifs_clusters)
```

And the actual plot:

```{r, fig.height = 6, fig.width = 12}
# Giant figure, suppl.
pfla_motif_clusters_minim <- 
  aggregate(
    pfla_motifs_clusters[,c(8,11,15)],
    by = list(
      motif = pfla_motifs_clusters$motif_minim,
      category = pfla_motifs_clusters$category
      ),
    FUN = max
    )

# provide ggplot a new ordering of the factor
pfla_motif_clusters_minim <- 
  pfla_motif_clusters_minim[
    order(
      pfla_motif_clusters_minim$category,
      pfla_motif_clusters_minim$motif
      ),
    ]

pfla_motif_clusters_minim$motif <-
  factor(
    pfla_motif_clusters_minim$motif,
    levels=unique(pfla_motif_clusters_minim$motif)
    )

ggplot(data = pfla_motif_clusters_minim, aes(x = motif, y = category, # esto ta der carajo
                                              color = `logqval`, size = as.factor(size_pct))) +
  geom_point() +
  scale_color_continuous_sequential(palette = "SunsetDark")+
  # scale_color_gradient(low = "#f1c232", high = "#5b0d80") +
  theme_bw() +
  ylab("") +
  xlab("") +
  ggtitle("Motif enrichment analysis") +
  scale_y_discrete(limits=rev) +
  theme(text = element_text(size=10), legend.text = element_text(size=10), axis.text.x = element_text(angle = 90, vjust = 0.5)) +
  scale_size_manual(values = c(1, 2, 3, 4,5), labels = c("1-10", "10-20", "20-40", "40+")) +
  guides(size=guide_legend(title="% peaks with motif")
  )
```

A summarised figure with motif enrichments pooled at the class level (perhaps less informative than the other one).

```{r}
pfla_motif_clusters_minim_perclass <- 
  aggregate_motifs_table(pfla_motifs_clusters)

pfla_motif_clusters_minim_perclass <-
  pfla_motif_clusters_minim_perclass[
    order(
      pfla_motif_clusters_minim_perclass$category,
      pfla_motif_clusters_minim_perclass$class
    ),
  ]

pfla_motif_clusters_minim_perclass$class <-
  factor(
    pfla_motif_clusters_minim_perclass$class,
    levels = unique(pfla_motif_clusters_minim_perclass$class)
    )

# Simpler version at the class level
ggplot(data = pfla_motif_clusters_minim_perclass, aes(x = class, y = category,
                                                color = `logqval`, size = as.factor(size_pct))) +
  geom_point() +
  scale_color_continuous_sequential(palette = "SunsetDark")+
  # scale_color_gradient(low = "#f1c232", high = "#5b0d80") +
  theme_bw() +
  ylab("") +
  xlab("") +
  ggtitle("Motif enrichment analysis") +
  scale_y_discrete(limits=rev) +
  theme(text = element_text(size=10), legend.text = element_text(size=10), axis.text.x = element_text(angle = 90, vjust = 0.5)) +
  scale_size_manual(values = c(1, 2, 3, 4,5), labels = c("1-10", "10-20", "20-40", "40+")) +
  guides(size=guide_legend(title="% peaks with motif")
  )

#' Also missing: homer motif analysis of differentially regulated proximal vs distal, all diffreg peaks as background
```

## Saving the data

And finally we save the data:

```{r}
save(
  pfla_atac,
  pfla_atac_kmeans,
  pfla_atac_hm,
  pfla_motifs_clusters,
  pfla_motif_clusters_minim,
  pfla_motif_clusters_minim_perclass,
  file = "outputs/rda/atacseq_analysis.rda"
)
```
